Libraries/Arduino_BNO055/BNO055_IMU.cpp

```diff
@@ -0,0 +1,138 @@
+#include <BNO055_IMU.h>
+
+void read_data(AxisData *data, DataMode mode)
+{
+    switch (mode) {
+        case GYRO:
+            {
+                uint8_t raw_data[6];
+                i2c_read(I2C_ADDRESS, GYR_DATA_X_LSB, 6, raw_data);
+                data->gyroAxes[0] = ((int16_t)raw_data[1] << 8) | raw_data[0];
+                data->gyroAxes[1] = ((int16_t)raw_data[3] << 8) | raw_data[2];
+                data->gyroAxes[2] = ((int16_t)raw_data[5] << 8) | raw_data[4];
+            }
+            break;
+        case ACCEL:
+            {
+                uint8_t raw_data[6];
+                i2c_read(I2C_ADDRESS, LIA_DATA_X_LSB, 6, raw_data);
+                data->accelAxes[0] = ((int16_t)raw_data[1] << 8) | raw_data[0];
+                data->accelAxes[1] = ((int16_t)raw_data[3] << 8) | raw_data[2];
+                data->accelAxes[2] = ((int16_t)raw_data[5] << 8) | raw_data[4];
+            }
+            break;
+        case GRAVITY:
+            {
+                uint8_t raw_data[6];
+                i2c_read(I2C_ADDRESS, GRV_DATA_X_LSB, 6, raw_data);
+                data->gravAxes[0] = ((int16_t)raw_data[1] << 8) | raw_data[0];
+                data->gravAxes[1] = ((int16_t)raw_data[3] << 8) | raw_data[2];
+                data->gravAxes[2] = ((int16_t)raw_data[5] << 8) | raw_data[4];
+            }
+            break;
+        case EUL:
+            {
+                uint8_t raw_data[6];
+                i2c_read(I2C_ADDRESS, EUL_HEADING_LSB, 6, raw_data);
+                data->eulerAngles[0] = ((int16_t)raw_data[1] << 8) | raw_data[0];
+                data->eulerAngles[1] = ((int16_t)raw_data[3] << 8) | raw_data[2];
+                data->eulerAngles[2] = ((int16_t)raw_data[5] << 8) | raw_data[4];
+            }
+            break;
+        case QUAT:
+            {
+                uint8_t raw_data[8];
+                i2c_read(I2C_ADDRESS, QUA_DATA_W_LSB, 8, raw_data);
+                data->quaternions[0] = ((int16_t)raw_data[1] << 8) | raw_data[0];
+                data->quaternions[1] = ((int16_t)raw_data[3] << 8) | raw_data[2];
+                data->quaternions[2] = ((int16_t)raw_data[5] << 8) | raw_data[4];
+                data->quaternions[3] = ((int16_t)raw_data[7] << 8) | raw_data[6];
+            }
+            break;
+    }
+}
+
+void update_data(Data *my_data)
+{
+    AxisData *data;
```

Comment:
Just create this on the stack, no need to dynamically allocate.

Reply:
I don't think you addressed this but it's more of a nitpick. I'll approve, your choice if you want to change it or merge as-is

Reply:
Ah, you're right. I did miss this one. I'll update that in a bit.